### PR TITLE
­[2.7] Properly detect if PodSecurityPolicies are enabled in K3s clusters during project creation

### PR DESCRIPTION
**Pull request by sbstp**
## Problem

Rancher does not properly detect if PSPs are enabled in a K3s cluster. This prevents creating a project with a specific PSP within the cluster.
 
## Solution

Check the `kube-apiserver-arg`s to see if PSPs are enabled for the cluster.
 
## Testing

The repro steps are mostly valid, although I did have to use a different guide than the one linked in the issue. I have synthesized the steps required to setup a k3s hardened cluster with PSPs enabled in this [repo](https://github.com/sbstp-suse/x-notes/blob/main/k3s-hardened/) and there's an `install.sh` script in there to perform all the steps necessary to setup hardening. Once the machine is ready, you can create a custom cluster using a spec similar to the one in the `README` file and register the cluster by running the command provided by rancher.

We're discussing the possibility of having a corral package to do make all of this easier, but it isn't ready yet.

### Manual Testing

1. Setup a hardened cluster on Digital Ocean droplet using the steps described above.
2. Create a new custom cluster and register it on the droplet.
3. Try to create a project with PSP set to `unrestricted`.

### Automated Testing

None

## QA Testing Considerations

None
 
### Regressions Considerations

None

### CODE REVIEW

File: pkg/api/norman/customization/project/project_actions.go [hunk 1]
```diff
@@ -52,6 +55,7 @@ type Handler struct {
 	ProjectLister     v3.ProjectLister
 	ClusterManager    *clustermanager.Manager
 	ClusterLister     v3.ClusterLister
+	ClusterListerEx   ctrlv1.ClusterController
```
Comment (sbstp) [inline pkg/api/norman/customization/project/project_actions.go:58]: This member needs a better name

Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:58]: Open to suggestions/standards from other places in the codebase

Comment (HarrisonWAffel) [reply pkg/api/norman/customization/project/project_actions.go:58]: Well, you aren't really using a `Lister` here, so maybe something more like `ProvisioningClusterController` or `ProvisioningClusterClient`? Also, have  you tried to see if you can use a cache for this? would be more efficient than a controller/client.

File: pkg/api/norman/customization/project/project_actions.go [hunk 2]
```diff
@@ -9,24 +9,27 @@ import (
...
 	"github.com/rancher/norman/httperror"
 	"github.com/rancher/norman/types"
 	"github.com/rancher/norman/types/convert"
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	v1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
 	client "github.com/rancher/rancher/pkg/client/generated/management/v3"
 	"github.com/rancher/rancher/pkg/clustermanager"
+	"github.com/rancher/rancher/pkg/controllers/management/imported"
 	"github.com/rancher/rancher/pkg/generated/compose"
+	ctrlv1 "github.com/rancher/rancher/pkg/generated/controllers/provisioning.cattle.io/v1"
```
Comment (HarrisonWAffel) [inline pkg/api/norman/customization/project/project_actions.go:24]: A lot of other places use `provisioningcontrollers` or `provcontrollers` when importing this package, or even `provisioningv1`

File: pkg/api/norman/customization/project/project_actions.go [hunk 3]
```diff
@@ -273,23 +277,28 @@ func (h *Handler) setPodSecurityPolicyTemplate(actionName string, action *types.
...
-		cluster, err := h.ClusterLister.Get("", clusterName)
+		clusterv3, err := h.ClusterLister.Get("", clusterName)
+		if err != nil {
+			return fmt.Errorf("error retrieving clusterv3 [%s]: %v", clusterName, err)
+		}
+
+		clusterv1, err := h.ClusterListerEx.Get("fleet-default", clusterv3.Spec.DisplayName, metav1.GetOptions{})
 		if err != nil {
-			return fmt.Errorf("error retrieving cluster [%s]: %v", clusterName, err)
+			return fmt.Errorf("error retrieving clusterv1 [%s]: %w", clusterv3.Spec.DisplayName, err)
```
Comment (HarrisonWAffel) [inline pkg/api/norman/customization/project/project_actions.go:295]: We should try to refer to `v1` cluster as the management cluster and the `v3` cluster as the provisioning cluster so it's clearer to other's who don't know the differentiation between the two versions

File: pkg/api/norman/customization/project/project_actions.go [hunk 4]
```diff
@@ -450,3 +459,32 @@ func getID(id interface{}) (string, error) {
...
+// are ignored.
+func parseKubeAPIServerArgs(cluster *v1.Cluster) map[string]string {
+	result := map[string]string{}
+	args := cluster.Spec.RKEConfig.MachineGlobalConfig.Data["kube-apiserver-arg"].([]any)
+	for _, arg := range args {
+		key, value, found := strings.Cut(arg.(string), "=")
+		if found {
+			result[key] = value
+		} else {
+			log.Warnf(
```
Comment (HarrisonWAffel) [inline pkg/api/norman/customization/project/project_actions.go:483]: Are there other areas of the norman package that log? Just curious, haven't looked into it too much

**Review by HarrisonWAffel — CHANGES_REQUESTED** [review]

File: pkg/api/norman/customization/project/project_actions.go (hunk 4 above)
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:483]: No there are no other instances. Maybe it's best to do something else

File: pkg/api/norman/customization/project/project_actions.go (hunk 3 above)
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:295]: names updated

File: pkg/api/norman/customization/project/project_actions.go (hunk 2 above)
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:24]: renamed to `provisioningv1`

File: pkg/api/norman/customization/project/project_actions.go (hunk 1 above)
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:58]: Renamed and used a cache.

File: pkg/api/norman/customization/project/project_actions.go (hunk 4 above)
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go]: I decided to just remove it and ignore invalid args if any.

**Review by snasovich — CHANGES_REQUESTED** [review]

File: pkg/api/norman/customization/project/project_actions.go [hunk 5]
```diff
@@ -450,3 +458,27 @@ func getID(id interface{}) (string, error) {
...
+	args := parseKubeAPIServerArgs(provisioningCluster)
+	return strings.Contains(args["enable-admission-plugins"], "PodSecurityPolicy")
+}
+
+// parseKubeApiServerArgs parses the "kube-apiserver-arg" available in the
+// clusters' MachineGlobalConfig to a map. The arguments are expected to
+// follow the "key=value" format. Arguments that don't follow this format
+// are ignored.
+func parseKubeAPIServerArgs(provisioningCluster *provisioningv1.Cluster) map[string]string {
+	result := map[string]string{}
```
[the remark this answers is not in the document]
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:475]: done

File: pkg/api/norman/customization/project/project_actions.go [hunk 6]
```diff
@@ -450,3 +458,27 @@ func getID(id interface{}) (string, error) {
...
+	args := provisioningCluster.Spec.RKEConfig.MachineGlobalConfig.Data["kube-apiserver-arg"].([]any)
```
[the remark this answers is not in the document]
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:476]: added checks

File: pkg/api/norman/customization/project/project_actions.go [hunk 7]
```diff
@@ -281,13 +284,18 @@ func (h *Handler) setPodSecurityPolicyTemplate(actionName string, action *types.
 		}
 
 		clusterName := idParts[0]
-		cluster, err := h.ClusterLister.Get("", clusterName)
+		managementCluster, err := h.ClusterLister.Get("", clusterName)
+		if err != nil {
+			return fmt.Errorf("error retrieving management cluster [%s]: %v", clusterName, err)
+		}
+
+		provisioningCluster, err := h.ProvisioningClusterClient.Get("fleet-default", managementCluster.Spec.DisplayName, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go]: added a check to ignore this, but as noted above by Harrison, there's no logging in norman

File: pkg/api/norman/customization/project/project_actions.go [hunk 8]
```diff
@@ -459,23 +476,29 @@ func isProvisionedRke2Cluster(cluster *v3.Cluster) bool {
...
+
+	args, ok := rawArgs.([]any)
+	if !ok || args == nil {
+		return nil
+	}
+
 	for _, arg := range args {
-		key, value, found := strings.Cut(arg.(string), "=")
+		s, ok := arg.(string)
+		if !ok {
```
Comment (jakefhyde) [inline pkg/api/norman/customization/project/project_actions.go:499]: Is this possible? If not, should we log an error here?

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/api/norman/customization/project/project_actions.go (hunk 8 above)
Comment (HarrisonWAffel) [reply pkg/api/norman/customization/project/project_actions.go:499]: @jakefhyde He initially had a log statement here which I pointed out. The `project` package doesn't seem to have any logging, so I wasn't sure if we wanted to add in a log for this scenario alone, or if a logger needed any prerequisite setup. 

If we do want to add logging for this, then my bad Simon!

Comment (jakefhyde) [reply pkg/api/norman/customization/project/project_actions.go:499]: You bring up a good point, if there's no logging in this package yet we are probably fine to just omit it.

File: pkg/api/norman/customization/project/project_actions.go [hunk 9]
```diff
@@ -444,6 +443,24 @@ func (h *Handler) updateBinding(binding map[string]interface{}, request *types.A
 	return nil
 }
 
+func (h *Handler) areK3SPodSecurityPoliciesEnabled(managementCluster *v3.Cluster) (bool, error) {
```
[the remark this answers is not in the document]
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go]: Is the stuff in `apis` generated by steve?

File: pkg/api/norman/customization/project/project_actions_test.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	provisioningv1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
+	rkev1 "github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1"
+)
+
+func TestParseKubeAPIServerArgs(t *testing.T) {
```
Comment (jakefhyde) [inline pkg/api/norman/customization/project/project_actions_test.go:12]: Nice, looks like we caught some additional cases too so that's good.

**Review by jakefhyde — DISMISSED** [review]

**Comment by sbstp** [comment]
I plan on squashing all these commits before merging. Once everything is approved.

**Review by HarrisonWAffel — DISMISSED** [review]

File: pkg/api/norman/customization/project/project_actions.go [hunk 11]
```diff
@@ -281,13 +284,18 @@ func (h *Handler) setPodSecurityPolicyTemplate(actionName string, action *types.
 		}
 
 		clusterName := idParts[0]
-		cluster, err := h.ClusterLister.Get("", clusterName)
+		managementCluster, err := h.ClusterLister.Get("", clusterName)
 		if err != nil {
-			return fmt.Errorf("error retrieving cluster [%s]: %v", clusterName, err)
+			return fmt.Errorf("error retrieving management cluster [%s]: %v", clusterName, err)
```
Comment (snasovich) [inline pkg/api/norman/customization/project/project_actions.go:289]: Noticed that we use `%w` in another place here to wrap errors, but not here or a couple of lines below. Fine to leave it not addressed, but ideally would like to go with `%w` for any new/updated code for errors.

**Review by snasovich — APPROVED** [review]

File: pkg/api/norman/customization/project/project_actions.go (hunk 11 above)
Comment (sbstp) [reply pkg/api/norman/customization/project/project_actions.go:289]: I can fix that when I push the rebase

**Review by HarrisonWAffel — APPROVED** [review]

**Review by jakefhyde — APPROVED** [review]

**Comment by sbstp** [comment]
Thanks!

**Pull request merged**